r/R/dplyr-summarize.R

```diff
@@ -18,19 +18,24 @@
 
 # The following S3 methods are registered on load if dplyr is present
 
-summarise.arrow_dplyr_query <- function(.data, ...) {
-  call <- match.call()
-  .data <- arrow_dplyr_query(.data)
-  if (query_on_dataset(.data)) {
-    not_implemented_for_dataset("summarize()")
+summarise.arrow_dplyr_query <- function(.data, ..., .engine = c("arrow", "duckdb")) {
+  if (match.arg(.engine) == "duckdb") {
+    tbl <- to_duckdb(.data)
+    dplyr::summarise(tbl, ...)
+  } else {
+    call <- match.call()
+    .data <- arrow_dplyr_query(.data)
+    if (query_on_dataset(.data)) {
+      not_implemented_for_dataset("summarize()")
+    }
+    exprs <- quos(...)
+    # Only retain the columns we need to do our aggregations
```

Comment:
Why not also retain only the columns we need to do aggregations when `.engine == "duckdb"`?

Reply:
Good catch, let me re-arrange the if/else to do most of the work / this above it.

Reply:
Good call, the way the duckdb implementation currently works, there's no pushdown of selection into the Arrow query from the duckdb side, so this optimization is something we should handle ourselves.

Reply:
Yeah, the pushdown will be here very shortly — I got an email from Pedro that he has a working version on a branch, though I haven't had a chance to test it myself yet. Regardless, I can't think of a reason _not_ to do this here anyway.

Reply:
I already have something that works, and I think we'll conflict anyway, so I don't think there's any harm keeping it in here?